### ISSUE TITLE
supposed compilation errors when using java 17 modules in visual studio code

### ISSUE THREAD

**Issue by bumby123** [redhat-developer/vscode-java#3376]
##### Environment
- Operating System: windows 11
- JDK version: "17.0.7" 2023-04-18 LTS 
- Visual Studio Code version: 1.84.0 user setup
- Java extension version: v.0.25.15

> When I create the following code I get a compilation error saying that certain packages do not exist. The steps I followed are:
> 
>     1. create a project in java in my case it is called OCP JAVA 17
> 
>     2. create a folder called 12. Modules (within this folder is the directory with program codes and modules written in Java 17)
> 
>     3. create a folder called feeding
> 
>     4. inside the feeding folder, create a module-info.java file with the following code
> 
> 
> ```java
> module zoo.animal.feeding {
>     exports zoo.animal.feeding;
> }
> ```
> 
>     5. inside the feeding folder, create a package or folders in the following order zoo/animal/feeding and add the Task.java file with the following code
> 
> 
> ```java
> package zoo.animal.feeding;
> 
> public class Task {
>     public static void main(String... args) {
>         System.out.println("All fed!");
>     }
> }
> ```
> 
>     6. compile Task.java class and module-info.java with following command in terminal opening from 12.Modules folder
>        javac -p
>        mods -d
>        feeding
>        feeding/zoo/animal/feeding/Task.java feeding/*.java
> 
>     7. create a folder called mods and run the following command in the terminal to package our compiled code from the 12.Modules folder
>        jar -cvf
>        mods/zoo.animal.feeding.jar -C
>        feeding/ .
> 
>     8. run the program by entering the following command in the terminal from 12.Modules
>        java -p
>        mods
>        -m
>        zoo.animal.feeding/zoo.animal.feeding.Task
>        must be printed in console All fed!
> 
>     9. create a folder called care with the following file module-info.java with the code
> 
> 
> ```java
> module zoo.animal.care {
>     exports zoo.animal.care.medical; //here is vsc supposed compilation error
> 
>     requires zoo.animal.feeding;
> }
> ```
> 
>     10. inside the care folder create 2 packages/folders in the following order zoo/animal/care/details and zoo/animal/care/medical
> 
>     11. in the zoo/animal/care/details package add the following file HippoBirthday.java
> 
> 
> ```java
> // HippoBirthday.java
> package zoo.animal.care.details;
> 
> import zoo.animal.feeding.*;
> 
> public class HippoBirthday {
>     private Task task;
>     public static void main(String... args) { //mio
>         System.out.println("Feliz cumple"); //mio
>     }
> }
> ```
> 
>     12. in the zoo/animal/care/medical package add the following Diet.java file
> 
> 
> ```java
> // Diet.java
> package zoo.animal.care.medical;
> 
> public class Diet {
>     public static void main(String... args) { //mio
>         System.out.println("Diet"); //mio
>     }
> }
> ```
> 
>     13. compile and package what is inside the care folder, in the terminal from 12.Modules with the following command
>         javac -p
>         mods
>         -d
>         care
>         care/zoo/animal/care/details/_.java
>         care/zoo/animal/care/medical/_.java
>         care/module-info.
>         java
> 
> 
> jar -cvf mods/zoo.animal.care.jar -C care/ . 14) run the classes that are inside the zoo.animal.care.medical package (which vsc says does not exist) in terminal from 12.Modules java -p mods -m zoo.animal.care/zoo.animal.care.medical.Diet the word Diet should be printed in the console
> 
> If the program runs properly, there should be no reason for vsc to mark an error in red, such as in the module-info.java file in the care folder that says that the package zoo.animal.care.medical does not exist. Your help please
> 
> attached screenshots
> 
> ![imagen](https://user-images.githubusercontent.com/85321074/280469248-4d4bf786-a73b-4b5c-8bf6-ae57750bb935.png) ![imagen](https://user-images.githubusercontent.com/85321074/280469264-824eb797-1ebe-4d2a-8f3b-9c881ba95a71.png)
> 
> VS Code version: Code 1.84.0 ([d037ac0](https://github.com/microsoft/vscode/commit/d037ac076cee195194f93ce6fe2bdfe2969cc82d), 2023-11-01T11:29:04.398Z) OS version: Windows_NT x64 10.0.22621 Modes:
> System Info
> Extensions (15)
> A/B Experiments

**Comment by maxandersen** [comment]
I find it hard to follow your example - can you attach a zip with a reproducer?

**Comment by bumby123** [comment]
Good morning Max, thank you for your help, I will send you the link to the repository with the complete project and instructions to execute it.

**Comment by snjeza** [comment]
This is an upstream Eclipse issue.
The related issues:
- https://github.com/eclipse-jdt/eclipse.jdt.core/issues/1465
- https://github.com/redhat-developer/vscode-java/issues/2116
- https://github.com/eclipse-m2e/m2e-core/issues/173

A workaround: move the feeding directory to a separate project.

**Comment by bumby123** [comment]
Amm what is upstream Eclipse issue? There is not solution for vsc? Is that for each module-info. java file I would have create separate projects? Amm There will be some way to disable that false compilation error because in the end the code does work, it's just those red lines that cause confusion

**Comment by snjeza** [comment]
> Amm what is upstream Eclipse issue

See https://github.com/eclipse-jdt/eclipse.jdt.core/issues/1465

> There is not solution for vsc? Is that for each module-info. java file I would have create separate projects?

Right.

**Comment by bumby123** [comment]
I understand the truth, creating independent projects for each module-info seems a bit inappropriate. Maybe there is an option in the visual studio code to disable build error warnings only in specific project folders, perhaps by manipulating the settings or launch json?

**Comment by stephan-herrmann** [comment]
As I happened to find this issue just now, I tried the steps in the description using ecj instead of javac, and there's only one warning I see in all this:
```
----------
1. WARNING in xxx/modules-java-17-problem-vsc-master/care/zoo/animal/care/details/HippoBirthday.java (at line 7)
        private Task task;
                     ^^^^
The value of the field HippoBirthday.task is not used
----------
1 problem (1 warning)
```
Other than that compilation and running works flawlessly.

Apparently, this is not a problem of the compiler, but of how vscode configures the "project". Is a project in vscode directly mapped to what jdt calls a project? How does vscode invoke the compiler, from which directory, which options are being passed?
